In [2]:
import pandas as pd
print ("Bienvenue dans ce Projet 2 sur Stroke Prediction")

Bienvenue dans ce Projet 2 sur Stroke Prediction


In [3]:
# importer le dataset healthcare-dataset-stroke-data.csv à analyser
df = pd.read_csv("../data/healthcare-dataset-stroke-data.csv")
display(df.head())

,id,gender,age,hypertension,heart_disease,ever_married,work_type,Residence_type,avg_glucose_level,bmi,smoking_status,stroke
0,9046,Male,67.0,0,1,Yes,Private,Urban,228.69,36.6,formerly smoked,1
1,51676,Female,61.0,0,0,Yes,Self-employed,Rural,202.21,NaN,never smoked,1
2,31112,Male,80.0,0,1,Yes,Private,Rural,105.92,32.5,never smoked,1
3,60182,Female,49.0,0,0,Yes,Private,Urban,171.23,34.4,smokes,1
4,1665,Female,79.0,1,0,Yes,Self-employed,Rural,174.12,24.0,never smoked,1


In [ ]:
df.shape
# on a 5110 lignes et 12 colonnes

(5110, 12)

In [34]:
# les dimensions de notre dataset
print(f"Notre dataset est constitué des données de : {df.shape[0]} patients observés et {df.shape[1]} variables")

Notre dataset est constitué des données de : 5110 patients observés et 12 variables


In [12]:
# les colonnes de notre dataset
print(f"les colonnes sont : {list(df.columns)}")

les colonnes sont : ['id', 'gender', 'age', 'hypertension', 'heart_disease', 'ever_married', 'work_type', 'Residence_type', 'avg_glucose_level', 'bmi', 'smoking_status', 'stroke']


### Compréhension des variables
Variable	    Signification
id	            Identifiant patient
gender	        Sexe
age	            Âge
hypertension	Hypertension (0 = Non, 1 = Oui)
heart_disease	Maladie cardiaque (0 = Non, 1 = Oui)
ever_married	Déjà marié(e)
work_type	    Type d'emploi
Residence_type	Urbain ou Rural
avg_glucose_level	Glycémie moyenne
bmi	            imc: Indice de masse corporelle
smoking_status	Statut tabagique
stroke	        AVC (0 = Non, 1 = Oui)

In [ ]:
# infos de notre dataset
print(df.info())

<class 'pandas.DataFrame'>
RangeIndex: 5110 entries, 0 to 5109
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 5110 non-null   int64  
 1   gender             5110 non-null   str    
 2   age                5110 non-null   float64
 3   hypertension       5110 non-null   int64  
 4   heart_disease      5110 non-null   int64  
 5   ever_married       5110 non-null   str    
 6   work_type          5110 non-null   str    
 7   Residence_type     5110 non-null   str    
 8   avg_glucose_level  5110 non-null   float64
 9   bmi                4909 non-null   float64
 10  smoking_status     5110 non-null   str    
 11  stroke             5110 non-null   int64  
dtypes: float64(3), int64(4), str(5)
memory usage: 479.2 KB
None


Parmi nos colonnes nous avons des types float, int et str

In [21]:
# vérifier la présence des valeurs manquantes
print(df.isna().any())

print(f"######################")
print(f"######################")
print(f"######################")

print(df.isna().sum())


id                   False
gender               False
age                  False
hypertension         False
heart_disease        False
ever_married         False
work_type            False
Residence_type       False
avg_glucose_level    False
bmi                   True
smoking_status       False
stroke               False
dtype: bool
######################
######################
######################
id                     0
gender                 0
age                    0
hypertension           0
heart_disease          0
ever_married           0
work_type              0
Residence_type         0
avg_glucose_level      0
bmi                  201
smoking_status         0
stroke                 0
dtype: int64


seule la colonne bmi ou imc contient 201 valeurs manquantes. 

In [54]:
# calculons sa représentation en %
print(f" La colonne bmi contient {round(df["bmi"].isna().sum()/df.shape[0] *100, 2)} % de valeurs manquantes")

 La colonne bmi contient 3.93 % de valeurs manquantes


In [24]:
df.columns[df.isna().any()]

Index(['bmi'], dtype='str')

In [25]:
# Filtrer les lignes où la colonne 'bmi' contient les nan
df[df["bmi"].isna()]

,id,gender,age,hypertension,heart_disease,ever_married,work_type,Residence_type,avg_glucose_level,bmi,smoking_status,stroke
1,51676,Female,61.0,0,0,Yes,Self-employed,Rural,202.21,NaN,never smoked,1
8,27419,Female,59.0,0,0,Yes,Private,Rural,76.15,NaN,Unknown,1
13,8213,Male,78.0,0,1,Yes,Private,Urban,219.84,NaN,Unknown,1
19,25226,Male,57.0,0,1,No,Govt_job,Urban,217.08,NaN,Unknown,1
27,61843,Male,58.0,0,0,Yes,Private,Rural,189.84,NaN,Unknown,1
...,...,...,...,...,...,...,...,...,...,...,...,...
5039,42007,Male,41.0,0,0,No,Private,Rural,70.15,NaN,formerly smoked,0
5048,28788,Male,40.0,0,0,Yes,Private,Urban,191.15,NaN,smokes,0
5093,32235,Female,45.0,1,0,Yes,Govt_job,Rural,95.02,NaN,smokes,0
5099,7293,Male,40.0,0,0,Yes,Private,Rural,83.94,NaN,smokes,0


In [32]:
# vérifier les doublons
print(f" Notre dataset contient: {df.duplicated().sum()} doublons")

 Notre dataset contient: 0 doublons


In [57]:
# Observons les statistiques descriptives des variables
df.loc[:, df.columns !='id'].describe()

,age,hypertension,heart_disease,avg_glucose_level,bmi,stroke
count,5110.000000,5110.000000,5110.000000,5110.000000,4909.000000,5110.000000
mean,43.226614,0.097456,0.054012,106.147677,28.893237,0.048728
std,22.612647,0.296607,0.226063,45.283560,7.854067,0.215320
min,0.080000,0.000000,0.000000,55.120000,10.300000,0.000000
25%,25.000000,0.000000,0.000000,77.245000,23.500000,0.000000
50%,45.000000,0.000000,0.000000,91.885000,28.100000,0.000000
75%,61.000000,0.000000,0.000000,114.090000,33.100000,0.000000
max,82.000000,1.000000,1.000000,271.740000,97.600000,1.000000


Notons que la moyenne d'une variable binaire telle que stroke correspond donc à la proportion de 1.

### Age
Notre population est très hétérogène
min -> les nourissons : 0.08 
mean et median (50%)-> les adultes : 43.23 et 45 respectivement
max -> les personnes âgées : 82

75% de nos patients observés ne dépassent pas 61 ans.
25% ne dépassent pas 25 ans.

### Hypertension
mean -> 0.097456 : 9.74% des patients sont hypertendus.

### heart_disease
mean -> 0.054012 : 5.4 % des patients ont une maladie cardiaque.

### stroke : AVC 
mean -> 0.048728 : 4.87 % des patients ont subi un AVC. Donc environ 95% des patients de notre dataset n'ont pas fait d'AVC.

### avg_glucose_level	
mean -> 55.12 
75% ne dépassent pas 114.090
max -> 271.74

### bmi	
mean -> 28.89 
75% ne dépassent pas 33.10
max -> 97.6



### Répartition des patients suivant les variables

In [ ]:
# répartition AVC
df["stroke"].value_counts()

stroke
0    4861
1     249
Name: count, dtype: int64

In [76]:
round(df["stroke"].value_counts(normalize=True) *100,2)

stroke
0    95.13
1     4.87
Name: proportion, dtype: float64

Les patients ayant subi un AVC représentent seulement 4,9 % de l'échantillon. Le jeu de données est donc fortement déséquilibré en faveur des patients n'ayant pas présenté d'AVC.

In [61]:
# répartition du sexe
df["gender"].value_counts()

gender
Female    2994
Male      2115
Other        1
Name: count, dtype: int64

In [75]:
round(df["gender"].value_counts(normalize=True) *100,2)

gender
Female    58.59
Male      41.39
Other      0.02
Name: proportion, dtype: float64

Notre dataset contient :

58,59 % de femmes
41,39 % d'hommes
1 observation "Other" qui représente 0.02%

In [62]:
# répartition tabagisme
df["smoking_status"].value_counts()

smoking_status
never smoked       1892
Unknown            1544
formerly smoked     885
smokes              789
Name: count, dtype: int64

In [74]:
round(df["smoking_status"].value_counts(normalize=True) *100,2)

smoking_status
never smoked       37.03
Unknown            30.22
formerly smoked    17.32
smokes             15.44
Name: proportion, dtype: float64

30 % de la donnée "Unknown" est présente dans la colonne tabagisme . Ce n'est pas une valeur manquante technique, mais une information manquante métier.

In [63]:
# répartition type d'emploi
df["work_type"].value_counts()

work_type
Private          2925
Self-employed     819
children          687
Govt_job          657
Never_worked       22
Name: count, dtype: int64

In [78]:
round(df["work_type"].value_counts(normalize=True)*100,2)

work_type
Private          57.24
Self-employed    16.03
children         13.44
Govt_job         12.86
Never_worked      0.43
Name: proportion, dtype: float64

La majorité travaille dans le secteur privé.
on a 22 patients soit (0.43%) qui n'ont jamais travaillé.


In [64]:
# répartition type de résidence
df["Residence_type"].value_counts()

Residence_type
Urban    2596
Rural    2514
Name: count, dtype: int64

In [80]:
round(df["Residence_type"].value_counts(normalize=True)*100,2)

Residence_type
Urban    50.8
Rural    49.2
Name: proportion, dtype: float64

Répartition très équilibrée :

Urbain : 50.8 %
Rural : 49.2 %

In [79]:
# Répartition Hypertention
round(df["hypertension"].value_counts(normalize=True)*100,2)

hypertension
0    90.25
1     9.75
Name: proportion, dtype: float64

In [66]:
# la moyenne d'age selon AVC
df.groupby("age")["stroke"].mean()

age
0.08     0.000000
0.16     0.000000
0.24     0.000000
0.32     0.000000
0.40     0.000000
           ...   
78.00    0.205882
79.00    0.200000
80.00    0.242857
81.00    0.233333
82.00    0.160714
Name: stroke, Length: 104, dtype: float64

Le résultat n'est pas très lisible et donc pas exploitable.
Nous allons à la place vérifier si les patients ayant subi un AVC sont plus âgés que ceux n'en ayant pas subi.


In [84]:
# Nous allons à la place vérifier si les patients ayant subi un AVC sont plus âgés que ceux n'en ayant pas subi
df.groupby("stroke")["age"].mean()

stroke
0    41.971545
1    67.728193
Name: age, dtype: float64

On constate que les patients ayant subi un AVC présentent un âge moyen de 67,7 ans contre seulement 42 ans pour les patients n'ayant pas subi d'AVC.
L'écart observé d'environ 26 ans 

In [ ]:
# Observer le nombre de patients par sexxe qui sont fait un AVC 
df.groupby("gender")["stroke"].mean()

gender
Female    0.047094
Male      0.051064
Other     0.000000
Name: stroke, dtype: float64

4.7 % des femmes ont eu un AVC
5.1 % des hommes ont eu un AVC

Les hommes présentent un taux d'AVC légèrement supérieur à celui des femmes dans ce dataset.
L'écart reste faible.

Pour l'instant ce n'est qu'une observation. Nous ne pouvons pas encore conclure que le sexe est un facteur majeur.

### Nous allons essayer de voir si les patients hypertendus ont plus fait l'AVC et de même pour les patients atteints de maladie cardiaque via une analyse exploratoire pour étudier la relation entre deux variables catégorielles.

In [97]:
# AVC selon hypertension
pd.crosstab(df["hypertension"], df["stroke"], normalize='index').round(3)

stroke,0,1
hypertension,,
0,0.960,0.040
1,0.867,0.133


En comparant le taux d'AVC nous avons :

Non hypertendus : 4,0 %
Hypertendus : 13,3 %
Les patients hypertendus présentent donc un risque d'AVC.

In [ ]:
# la différence entre les 2 taux
print(f"Les patients hypertendus présentent donc un risque d'AVC : {13.3/4} fois plus élévés") 

Les patients hypertendus présentent donc un risque d'AVC : 3.325 fois plus élévés


Les patients souffrant d'hypertension présentent une proportion d'AVC nettement supérieure à celle observée chez les patients non hypertendus (13,3 % contre 4,0 %). 
L'hypertension apparaît ainsi comme un facteur fortement associé à la survenue d'un AVC.

In [99]:
# AVC selon maladie Cradiaque 
pd.crosstab(df["heart_disease"], df["stroke"], normalize='index').round(3)

stroke,0,1
heart_disease,,
0,0.958,0.042
1,0.830,0.170


En comparant le taux d'AVC :

Sans maladie cardiaque : 4,2 %
Avec maladie cardiaque : 17,0 %
Les patients atteints d'une maladie cardiaque présentent donc un risque d'AVC.

In [115]:
# la différence entre les 2 taux
print(f"Les patients atteints d'une maladie cardiaque présentent donc un risque d'AVC : {(17.0/4.2)} fois plus élévés") 

Les patients atteints d'une maladie cardiaque présentent donc un risque d'AVC : 4.0476190476190474 fois plus élévés


Les patients atteints d'une maladie cardiaque présentent un taux d'AVC d'environ 17 %, contre seulement 4,2 % chez les patients sans antécédent cardiaque. 
La présence d'une maladie cardiaque semble donc fortement associée au risque d'AVC.

In [118]:
# glycémie par AVC
df.groupby("stroke")["avg_glucose_level"].mean().round(3)

stroke
0    104.796
1    132.545
Name: avg_glucose_level, dtype: float64

Les patients ayant subi un AVC présentent une glycémie moyenne significativement plus élevée : 132,5 contre 104,8.
La glycémie apparaît comme un facteur potentiellement associé au risque d'AVC.

In [121]:
# IMC par AVC
df.groupby("stroke")["bmi"].mean().round(3)

stroke
0    28.823
1    30.471
Name: bmi, dtype: float64

L'IMC moyen chez les patients ayant subi un AVC est légèrement supérieur à celui des autres patients. 30,5 contre 28,8.
Une association semble exister mais apparaît moins marquée.


In [123]:
# Tabagisme par AVC
pd.crosstab(df["smoking_status"], df["stroke"], normalize='index').round(3)

stroke,0,1
smoking_status,,
Unknown,0.970,0.030
formerly smoked,0.921,0.079
never smoked,0.952,0.048
smokes,0.947,0.053


Le taux le plus élevé est observé chez Formerly smoked avec 7,9 % contre 4.8 % chea les non-fumeurs.
On observe une association entre fumer et AVC mais il y'a certainement d'autres facteurs liés pour parler de causalité.